File: openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/Box2.java
package io.lacuna.artifex;

import java.util.ArrayList;
import java.util.List;

import static io.lacuna.artifex.Vec.vec;
import static java.lang.Double.NaN;

/**
 * @author ztellman
 */
public class Box2 extends Box<Vec2, Box2> {

  public static final Box2 EMPTY = new Box2(vec(NaN, NaN), vec(NaN, NaN));

  public final double lx, ly, ux, uy;

  Box2(double ax, double ay, double bx, double by) {
    if (ax < bx) {
      this.lx = ax;
      this.ux = bx;
    } else {
      this.ux = ax;
      this.lx = bx;
    }

    if (ay < by) {
      this.ly = ay;
      this.uy = by;
    } else {
      this.uy = ay;
      this.ly = by;
    }
  }

  public Box3 box3(double lz, double uz) {
    return new Box3(lx, ly, lz, ux, uy, uz);
  }

  public double width() {
    return ux - lx;
  }

  public double height() {
    return uy - ly;
  }

  public Box2(Vec2 a, Vec2 b) {
    this(a.x, a.y, b.x, b.y);
  }

  public Box2 scale(double k) {
    return scale(vec(k, k));
  }

  public Box2 scale(double x, double y) {
    return scale(vec(x, y));
  }

  public Box2 translate(double x, double y) {
    return translate(vec(x, y));
  }

  public Vec2[] vertices() {
    return new Vec2[] {vec(lx, ly), vec(ux, ly), vec(ux, uy), vec(lx, uy)};
  }

  public Ring2 outline() {
    List<Curve2> cs = new ArrayList<>();
    Vec2[] vs = vertices();
    for (int i = 0; i < vs.length; i++) {
      cs.add(Line2.line(vs[i], vs[(i + 1) % 4]));
    }
    return new Ring2(cs);
  }

  @Override
  public boolean intersects(Box2 b) {
    if (isEmpty() || b.isEmpty()) {
      return false;
    }

    return b.ux >= lx
      & ux >= b.lx
      & b.uy >= ly
      & uy >= b.ly;
  }

  @Override
  public Vec2 lower() {
    return new Vec2(lx, ly);
  }

  @Override
  public Vec2 upper() {
    return new Vec2(ux, uy);
  }

  @Override
  public boolean isEmpty() {
    return this == EMPTY;
  }

  @Override
  protected Box2 construct(Vec2 a, Vec2 b) {
    return new Box2(a.x, a.y, b.x, b.y);
  }

  @Override
  protected Box2 empty() {
    return EMPTY;
  }
}

File: openrndr-kartifex/src/commonMain/kotlin/Box2.kt
package org.openrndr.kartifex


/**
 * @author ztellman
 */
class Box2 internal constructor(ax: Double, ay: Double, bx: Double, by: Double) :
    Box<Vec2, Box2>() {
    val lx: Double
    val ly: Double
    val ux: Double
    val uy: Double
    fun box3(lz: Double, uz: Double): Box3 {
        return Box3(lx, ly, lz, ux, uy, uz)
    }

    fun width(): Double {
        return ux - lx
    }

    fun height(): Double {
        return uy - ly
    }

    constructor(a: Vec2, b: Vec2) : this(a.x, a.y, b.x, b.y) {}

    fun scale(k: Double): Box2 {
        return scale(Vec2(k, k))
    }

    fun scale(x: Double, y: Double): Box2 {
        return scale(Vec2(x, y))
    }

    fun translate(x: Double, y: Double): Box2 {
        return translate(Vec2(x, y))
    }

    fun vertices(): Array<Vec2> {
        return arrayOf(
            Vec2(lx, ly),
            Vec2(ux, ly),
            Vec2(ux, uy),
            Vec2(lx, uy)
        )
    }

    fun outline(): Ring2 {
        val cs: MutableList<Curve2> = mutableListOf()
        val vs: Array<Vec2> = vertices()
        for (i in vs.indices) {
            cs.add(Line2.line(vs[i], vs[(i + 1) % 4]))
        }
        return Ring2(cs)
    }

    override fun intersects(b: Box2): Boolean {
        return if (isEmpty || b.isEmpty) {
            false
        } else (b.ux >= lx
                ) and (ux >= b.lx
                ) and (b.uy >= ly
                ) and (uy >= b.ly)
    }

    override fun lower(): Vec2 {
        return Vec2(lx, ly)
    }

    override fun upper(): Vec2 {
        return Vec2(ux, uy)
    }

    override val isEmpty: Boolean
        get() = this === EMPTY

    protected override fun construct(a: Vec2, b: Vec2): Box2 {
        return Box2(a.x, a.y, b.x, b.y)
    }

    protected override fun empty(): Box2 {
        return EMPTY
    }

    companion object {
        val EMPTY =
            Box2(Vec2(Double.NaN, Double.NaN), Vec2(Double.NaN, Double.NaN))
    }

    init {
        if (ax < bx) {
            lx = ax
            ux = bx
        } else {
            ux = ax
            lx = bx
        }
        if (ay < by) {
            ly = ay
            uy = by
        } else {
            uy = ay
            ly = by
        }
    }
}